test/Kestrel.FunctionalTests/Http2/H2SpecTests.cs

```diff
@@ -0,0 +1,135 @@
+                await host.StartAsync();
+
+                H2SpecCommands.RunTest(testCase.Id, host.GetPort(), testCase.Https, Logger);
+            }
+        }
+
+        public static TheoryData<H2SpecTestCase> H2SpecTestCases
+        {
+            get
+            {
+                var dataset = new TheoryData<H2SpecTestCase>();
+                var toSkip = new[] { "hpack/4.2/1", "http2/5.1/8", "http2/6.9.1/2", "http2/6.9.1/3", "http2/8.1.2.3/1",
+                    "http2/8.1.2.6/1", "http2/8.1.2.6/2" };
+
+                foreach (var testcase in H2SpecCommands.EnumerateTestCases())
+                {
+                    string skip = null;
+                    if (toSkip.Contains(testcase.Item1))
+                    {
+                        skip = "https://github.com/aspnet/KestrelHttpServer/issues/2154";
+                    }
+                    
+                    dataset.Add(new H2SpecTestCase()
+                    {
+                        Id = testcase.Item1,
+                        Description = testcase.Item2,
+                        Https = false,
+                        Skip = skip,
+                    });
+                    
+                    dataset.Add(new H2SpecTestCase()
+                    {
+                        Id = testcase.Item1,
+                        Description = testcase.Item2,
+                        Https = true,
+                        Skip = skip,
+                    });
+                }
+
+                return dataset;
+            }
+        }
+
+        public class H2SpecTestCase : IXunitSerializable
+        {
+            // For the serializer
+            public H2SpecTestCase()
+            {
+            }
+
+            public string Id { get; set; }
+            public string Description { get; set; }
+            public bool Https { get; set; }
+            public string Skip { get; set; }
+
+            public void Deserialize(IXunitSerializationInfo info)
+            {
+                Id = info.GetValue<string>(nameof(Id));
+                Description = info.GetValue<string>(nameof(Description));
+                Https = info.GetValue<bool>(nameof(Https));
+                Skip = info.GetValue<string>(nameof(Skip));
+            }
+
+            public void Serialize(IXunitSerializationInfo info)
+            {
+                info.AddValue(nameof(Id), Id, typeof(string));
+                info.AddValue(nameof(Description), Description, typeof(string));
+                info.AddValue(nameof(Https), Https, typeof(bool));
+                info.AddValue(nameof(Skip), Skip, typeof(string));
+            }
+
+            public override string ToString()
+            {
+                return $"{Id}, HTTPS:{Https}, {Description}";
+            }
+        }
+
+        private void ConfigureHelloWorld(IApplicationBuilder app)
+        {
+            app.Run(context =>
+            {
+                return context.Request.Body.CopyToAsync(context.Response.Body);
```

Comment:
Any specific reason for copying the request body rather than just responding with "Hello World"?

Reply:
Some of the tests deal with invalid request frame sequences. If you don't wait for the request body and immediately respond with Hello World then the client may not observe any errors and the test fails. Echoing the body was the most reliable way to wait for request errors.